### PR TITLE
Synchronize Javacp and Metac on several fronts

### PR DESCRIPTION
**Pull request by olafurpg**
* Unify module and class symbols for java classes in metac
* Move java static decls to java class decls in metac
* Disambiguate method symbols by unified java module/class symbol
* Escape `<init>` names in javacp
* Sort methods by `isStatic` in javacp

### CODE REVIEW

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala [hunk 1]
```diff
@@ -402,7 +411,10 @@ object Javacp {
   private def sarray(tpe: s.Type): s.Type =
     styperef("_root_.scala.Array#", tpe :: Nil)
 
-  private def styperef(symbol: String, args: List[s.Type] = Nil, prefix: Option[s.Type] = None): s.Type = {
+  private def styperef(
+      symbol: String,
+      args: List[s.Type] = Nil,
+      prefix: Option[s.Type] = None): s.Type = {
```
Comment (xeno-by) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: Why wasn't this autoformatted before?

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/DenotationOps.scala [hunk 2]
```diff
@@ -118,7 +120,9 @@ trait DenotationOps { self: DatabaseOps =>
 
     private def newInfo: (Option[s.Type], List[g.Symbol]) = {
       val ginfo = {
-        if (gsym.isGetter && gsym.isLazy && !gsym.isClass) {
+        if (gsym.isJavaClass) {
+          gsym.companionClass.info
+        } else if (gsym.isGetter && gsym.isLazy && !gsym.isClass) {
```
Comment (xeno-by) [inline semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/DenotationOps.scala]: Looks like the three `isJavaClass` checks above can be replaced by a single condition in `val gsym` above. I'll experiment with this myself after merging this PR.

File: tests/jvm/src/test/resources/highlevel.expect [hunk 3]
```diff
@@ -382,14 +382,14 @@ Names:
 [42..51): MetacJava <= _root_.example.MetacJava#
 [52..52):  <= _root_.example.MetacJava#`<init>`().
 [56..62): javacp => _root_.com.javacp.
-[63..72): MetacJava => _root_.com.javacp.MetacJava.
-[73..84): StaticInner => _root_.com.javacp.MetacJava.StaticInner.
-[85..93): isStatic => _root_.com.javacp.MetacJava.StaticInner.isStatic().
+[63..72): MetacJava => _root_.com.javacp.MetacJava#
+[73..84): StaticInner => _root_.com.javacp.MetacJava#StaticInner#
+[85..93): isStatic => _root_.com.javacp.MetacJava#StaticInner#isStatic().
```
Comment (xeno-by) [inline tests/jvm/src/test/resources/highlevel.expect]: 😍

File: tests/jvm/src/test/resources/highlevel.expect [hunk 4]
```diff
@@ -1251,8 +1264,8 @@ Symbols:
 _root_.java. => package java
 _root_.java.nio. => package nio
 _root_.java.nio.file. => package file
-_root_.java.nio.file.LinkOption. => final javadefined object LinkOption
-_root_.java.nio.file.LinkOption.NOFOLLOW_LINKS. => final javadefined val NOFOLLOW_LINKS: LinkOption
+_root_.java.nio.file.LinkOption# => abstract final sealed javadefined class LinkOption
```
Comment (xeno-by) [inline tests/jvm/src/test/resources/highlevel.expect]: `abstract` and `sealed` are interesting additions. `final` is pretty nuts. Looking forward to getting this sorted out in https://github.com/scalameta/scalameta/issues/1162.

File: tests/jvm/src/test/resources/lowlevel.expect [hunk 5]
```diff
@@ -536,23 +558,8 @@ _root_.com.javacp.MetacJava#overload(A+1). => def overload: (a: <?>): Unit
...
-_root_.com.javacp.MetacJava.StaticInner#NonStatic#method(NonStatic). => def method: (e: <?>): Unit
-  e => _root_.com.javacp.MetacJava.StaticInner#NonStatic#method(NonStatic).(e)
-  Unit => _root_.scala.Unit#
-_root_.com.javacp.MetacJava.StaticInner#`<init>`(). => primaryctor <init>: (): StaticInner
-  StaticInner => _root_.com.javacp.MetacJava.StaticInner#
-_root_.com.javacp.MetacJava.StaticInner#isNotStatic(). => def isNotStatic: (): Unit
-  Unit => _root_.scala.Unit#
-_root_.com.javacp.MetacJava.StaticInner. => final object StaticInner
-_root_.com.javacp.MetacJava.StaticInner.isStatic(). => def isStatic: (): Unit
+_root_.com.javacp.MetacJava#overload(A+3). => def overload: (a: <?>): Unit
```
Comment (xeno-by) [inline tests/jvm/src/test/resources/lowlevel.expect]: Why is this a `<?>`?

File: tests/jvm/src/test/resources/lowlevel.expect [hunk 6]
```diff
@@ -1240,9 +1255,12 @@ Symbols:
 _root_.java. => package java
 _root_.java.nio. => package nio
 _root_.java.nio.file. => package file
-_root_.java.nio.file.LinkOption. => final object LinkOption
-_root_.java.nio.file.LinkOption.NOFOLLOW_LINKS. => final val NOFOLLOW_LINKS: NOFOLLOW_LINKS.type
-  NOFOLLOW_LINKS => _root_.java.nio.file.LinkOption.NOFOLLOW_LINKS.
+_root_.java.nio.file.LinkOption# => abstract final sealed class LinkOption.{+4 decls}
+  extends Enum[LinkOption]
+  extends OpenOption
+  extends CopyOption
```
Comment (xeno-by) [inline tests/jvm/src/test/resources/lowlevel.expect]: 👍

File: tests/jvm/src/test/resources/metacp.expect [hunk 7]
```diff
@@ -310,50 +310,58 @@ Schema => SemanticDB v3
...
 Language => Java
-Symbols => 25 entries
+Symbols => 31 entries
 
 Symbols:
 _root_. => package _root_
 _root_.com. => package com
 _root_.com.javacp. => package javacp
-_root_.com.javacp.MetacJava# => class MetacJava.{+6 decls}
+_root_.com.javacp.MetacJava# => class MetacJava.{+8 decls}
```
Comment (xeno-by) [inline tests/jvm/src/test/resources/metacp.expect]: This is out of sync with javacp. Do you think that should be a blocker, or we can open a ticket for the future?

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala [hunk 8]
```diff
@@ -151,9 +157,13 @@ object Javacp {
           m.descriptor == method.descriptor
         }
         val suffix =
-          if (synonyms.length == 1) ""
+          if (synonyms.lengthCompare(1) == 0) ""
```
Comment (xeno-by) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: I think this is an overkill, but I don't think this is a blocker.

File: tests/jvm/src/test/resources/metac-metacp.expect.diff [hunk 9]
```diff
@@ -39,16 +39,20 @@ _root_.com.javacp.MetacJava#
...
 +++ metacp
-         symbol: "_root_.java.lang.Object#"
--    declarations: "_root_.com.javacp.MetacJava#`<init>`()."
--    declarations: "_root_.com.javacp.MetacJava#Overload1."
+     declarations: "_root_.com.javacp.MetacJava#`<init>`()."
+-    declarations: "_root_.com.javacp.MetacJava#Overload1#"
 -    declarations: "_root_.com.javacp.MetacJava#Overload1#"
--    declarations: "_root_.com.javacp.MetacJava#Overload2."
 -    declarations: "_root_.com.javacp.MetacJava#Overload2#"
-+    declarations: "_root_.com.javacp.MetacJava#<init>()."
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply tests/jvm/src/test/resources/metac-metacp.expect.diff]: I chickened out of reviewing this diff. 🐔

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala (hunk 1 above)
Comment (olafurpg) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: the `.java` part of `project.excludeFilters` ignored everything in the `javacp` directory. Fixed in https://github.com/scalameta/scalameta/pull/1417

Note to self, I should make this configuration easier in scalafmt. Or alternatively, we should let scalafmt loose on the scalameta repo! 😄

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/DenotationOps.scala (hunk 2 above)
Comment (olafurpg) [reply semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/DenotationOps.scala]: Nice job cleaning this up in #1416

File: tests/jvm/src/test/resources/highlevel.expect (hunk 4 above)
Comment (olafurpg) [reply tests/jvm/src/test/resources/highlevel.expect]: haha yes, `abstract final` makes no sense :joy:

File: tests/jvm/src/test/resources/lowlevel.expect (hunk 5 above)
Comment (olafurpg) [reply tests/jvm/src/test/resources/lowlevel.expect]: Not sure, looking at lowlevel.expect there seems to be quite a few other `<?>` which seem unrelated to this PR. I opened https://github.com/scalameta/scalameta/issues/1418 to track fixing these. I don't think this is a blocker at the moment.

File: tests/jvm/src/test/resources/metacp.expect (hunk 7 above)
Comment (olafurpg) [reply tests/jvm/src/test/resources/metacp.expect]: Good catch! Looking at the diff it seems metac emits too duplicate symbols for inner classes, which is a bug but not a blocker

```diff
============================
_root_.com.javacp.MetacJava#
============================
--- metac
+++ metacp
     declarations: "_root_.com.javacp.MetacJava#`<init>`()."
-    declarations: "_root_.com.javacp.MetacJava#Overload1#"
-    declarations: "_root_.com.javacp.MetacJava#Overload1#"
-    declarations: "_root_.com.javacp.MetacJava#Overload2#"
-    declarations: "_root_.com.javacp.MetacJava#Overload2#"
     declarations: "_root_.com.javacp.MetacJava#overload(A+1)."
     declarations: "_root_.com.javacp.MetacJava#overload(A+2)."
-    declarations: "_root_.com.javacp.MetacJava#StaticInner#"
-    declarations: "_root_.com.javacp.MetacJava#StaticInner#"
-    declarations: "_root_.com.javacp.MetacJava#Overload3#"
-    declarations: "_root_.com.javacp.MetacJava#Overload3#"
     declarations: "_root_.com.javacp.MetacJava#overload(A+3)."
+    declarations: "_root_.com.javacp.MetacJava#Overload2#"
+    declarations: "_root_.com.javacp.MetacJava#Overload3#"
+    declarations: "_root_.com.javacp.MetacJava#Overload1#"
+    declarations: "_root_.com.javacp.MetacJava#StaticInner#"
 accessibility {
```

Tracked in #1411

File: semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala (hunk 8 above)
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: I wouldn't say this is idiomatic, regardless of what Intellij thinks about it. `lengthCompare` is likely to be faster than `length` on big lists, but I don't think that performance gains are worth the readability hit here.

Comment (olafurpg) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala]: I like using a method designed exactly for this usecase. The readability hit is miniscule it doesn't bother me much.

**Pull request merged**